File: 278 First Bad Version.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        if (n == 1 || n == 0) return n;
        int lo = 1, up = n;
        while (lo < up) {
            int mid = lo + (up - lo) / 2;
            if (isBadVersion(mid)) up = mid;
            else lo = mid + 1;
        }
        return lo;
    }
}

File: 278 First Bad Version.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):

class Solution:
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        lo, hi = 1, n
        while lo <= hi:
            mi = (hi + lo) // 2
            if isBadVersion(mi):
                if mi == 1 or not isBadVersion(mi - 1):
                    return mi
                hi = mi - 1
            else:
                lo = mi + 1
        return None        
        